### PR TITLE
Modified lib/adagrams.rb and test/adagrams_test.rb

### PR DESCRIPTION
**Pull request by cescarez**
# Assignment Submission: Adagrams

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Feature | Feedback
--- | ---
What are the components that make up a method? | signature (name, parameters), `def` and `end` keywords, code block
What are the advantages of using git when collaboratively working on one code base? | Commit-tracking (time and author), the ability to revert back to clearly marked points, clear indications of version/merge conflicts, not needing to send code over email.
What kind of relationship did you and your pair have with the unit tests? | Unit tests brought clarity to limitations/bugs within code and allowed for a clear Red-Green-Refactor workflow. Automating/writing tests forces the consideration of edge cases up-front.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used `.map` when iterating through arrays and hashes and assigning accessed values to a new array.
What was one method you and your pair used to debug code? | The RubyMine debugger, `p` statements, running individual tests from within RubyMine.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Ask vs Guess culture, what each person needs/wants to be successful while paired together.

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10
+}
+
+def draw_letters
+  all_letters = ALL_LETTERS.map do |keys, values|
+    [keys.to_s] * values    #follow-up, why do square brackets work??
```
Comment (kaidamasaki) [inline lib/adagrams.rb:65]: Ruby pretty much extends the idea of `*` as repeated `+` to both Strings and Arrays.

So for an Array if you do: `[1, 2] * 3` you basically get `[1, 2] + [1, 2] + [1, 2]` or `[1, 2, 1, 2, 1, 2]`.

And for a String if you do `"Ada" * 3` you basically get `"Ada" + "Ada" + "Ada"` or `"AdaAdaAda"`.

The brackets around `keys.to_s` means that instead of having `"A" * 9` for the first iteration and getting `"AAAAAAAAA"` you have `["A"] * 9` for the first iteration and get `["A", "A", "A", "A", "A", "A", "A", "A", "A"]`.

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    U: 1,
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10
+}
+
+def draw_letters
+  all_letters = ALL_LETTERS.map do |keys, values|
```
Comment (kaidamasaki) [inline lib/adagrams.rb:64]: This only operates on a _single_ key and value at a time.

```suggestion
  all_letters = ALL_LETTERS.map do |key, value|
```

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    [keys.to_s] * values    #follow-up, why do square brackets work??
+  end
+
+  return all_letters.flatten.shuffle.first(10)
```
Comment (kaidamasaki) [inline lib/adagrams.rb:68]: You might want to look at `flat_map` for future assignments.  It's a `map` followed by a `flatten` all in one!

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+end
+
+def count_letters(letters_array)
+  letter_count = {}
+
+  letters_array.each do |letter|
+    (letter_count.include? letter) ? letter_count[letter] += 1 : letter_count[letter] = 1
+  end
+
+  return letter_count
```
Comment (kaidamasaki) [inline lib/adagrams.rb:78]: If you pass a default value to `Hash.new` you can simplify your code here:

```suggestion
  letter_count = Hash.new(0)

  letters_array.each do |letter|
    letter_count[letter] += 1
  end

  return letter_count
```

When you give `Hash.new` an argument it returns that value instead of `nil` when a key is missing.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+end
+
+def uses_available_letters?(input, letters_in_hand)
+  hand_count = count_letters(letters_in_hand)
+  input_count = count_letters(input.split(''))
```
Comment (kaidamasaki) [inline lib/adagrams.rb:83]: `chars` is a slightly clearer way to do this:

```suggestion
  input_count = count_letters(input.chars)
```

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      highest_score = { word: word, score: word_value }
+    elsif (word_value == highest_score[:score]) && (highest_score[:word].length != 10)
+      if (word.length == 10) || (word.length < highest_score[:word].length)
+        highest_score = { word: word, score: word_value }
+      end
+    end
+  end
+
+  return highest_score
+end
```
Comment (kaidamasaki) [inline lib/adagrams.rb:126]: This is a really elegant solution for this.

**Review by kaidamasaki — COMMENTED** [review]
Excellent job!  I've included a few small tips for improvement but this is a really clean and elegant solution!

# Adagrams



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Correctly creates and calls methods with proper syntax (parameters, return statements, etc.) | ✔️
| Uses correct syntax for conditional logic and iteration | ✔️
| Practices git with at least 3 small commits and meaningful commit messages | ✔️  You even had commits from **both** of your accounts!
| Utilizes unit tests to verify code; tests can run using the command `$ rake` and we see test successes and/or failures | ✔️
| Practices pair programming; the reflection question on pair programming is answered | ✔️

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| For the `draw_letters` method, there is an appropriate data structure to store the letter distribution. (You are more likely to draw an 'E' than an 'X'.) | ✔️
| Utilizes unit tests to verify code; all tests for `draw_letters` and `uses_available_letters?` pass | ✔️
| Utilizes unit tests to verify code; all tests for `score_word` pass | ✔️
| Utilizes unit tests to verify code; all tests for `highest_score_from` pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 4+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 3+ in Code Review && 2+ in Functional Requirements | 
| Red (Not at Standard) | 0-2 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
